search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,498 @@
+                    G_CALLBACK (handle_get_subsearch_result_set), self);
+    g_signal_connect (self->skeleton, "handle-get-result-metas",
+                    G_CALLBACK (handle_get_result_metas), self);
+    g_signal_connect (self->skeleton, "handle-activate-result",
+                    G_CALLBACK (handle_activate_result), self);
+    g_signal_connect (self->skeleton, "handle-launch-search",
+                    G_CALLBACK (handle_launch_search), self);
+
+    g_dbus_interface_skeleton_export (G_DBUS_INTERFACE_SKELETON (self->skeleton),
+                                      connection,
+                                      "/org/gnome/Yelp/SearchProvider",
+                                      NULL);
+    g_dbus_object_manager_server_set_connection (self->object_manager, connection);
+}
+
+static void
+search_provider_app_dispose (GObject *obj)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (obj);
+    if (self->name_owner_id != 0) {
+        g_bus_unown_name (self->name_owner_id);
+        self->name_owner_id = 0;
+    }
+
+    if (self->skeleton != NULL) {
+        g_dbus_interface_skeleton_unexport (G_DBUS_INTERFACE_SKELETON (self->skeleton));
+        g_clear_object (&self->skeleton);
+    }
+
+    g_clear_object (&self->object_manager);
+    g_clear_object (&self->doc);
+    g_clear_pointer (&self->page_data_hash_map, g_hash_table_unref);
+
+    G_OBJECT_CLASS (yelp_search_provider_app_parent_class)->dispose (obj);
+}
+
+static void
+preload_data_cb (YelpDocument          *document,
+                       YelpDocumentSignal    signal,
+                       YelpSearchProviderApp *app,
+                       GError                *error)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (app);
+    gchar **page_ids = yelp_document_list_page_ids (document);
+    gchar **iter;
+
+    for (iter = page_ids; *iter; iter++){
+        gchar *page_id = *iter;
+        PageData *data = g_new0 (PageData, 1);
+        gchar *icon_string = yelp_document_get_page_icon (document, page_id);
+        GIcon *icon = g_icon_new_for_string (icon_string, &error);
+        g_free (icon_string);
+
+        data = page_data_new_steal (yelp_document_get_page_title (document, page_id), 
+                                yelp_document_get_page_desc (document, page_id), 
+                                icon);
+        g_hash_table_insert (app->page_data_hash_map, page_id, data);
+    }
+
+    self->name_owner_id = g_bus_own_name (G_BUS_TYPE_SESSION,
+                                         "org.gnome.Yelp.SearchProvider",
+                                         G_BUS_NAME_OWNER_FLAGS_NONE,
+                                         search_provider_bus_acquired_cb,
+                                         search_provider_name_acquired_cb,
+                                         search_provider_name_lost_cb,
+                                         app, NULL);
+    g_free (page_ids);
```

Comment:
Maybe you could add a comment here that you stole the contents of `page_ids` so the only thing you free is the array that held the pointers to contents.